File: api/Services/TrainStateService.cs
using System.Collections.Concurrent;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using Api.Models;

namespace Api.Services;

public class TrainStateService
{
    private readonly ConcurrentDictionary<string, TrainPosition> _trains = new();
    private readonly ConcurrentDictionary<string, WebSocket> _clients = new();
    private readonly ILogger<TrainStateService> _logger;

    public TrainStateService(ILogger<TrainStateService> logger)
    {
        _logger = logger;
    }

    public void UpdateTrain(TrainPosition position)
    {
        _trains[position.Rid] = position;
        _ = BroadcastUpdateAsync(position);
    }

    public void RemoveTrain(string rid)
    {
        _trains.TryRemove(rid, out _);
        _ = BroadcastRemovalAsync(rid);
    }

    public async Task HandleClientAsync(WebSocket ws, CancellationToken ct)
    {
        var clientId = Guid.NewGuid().ToString();
        _clients[clientId] = ws;

        try
        {
            await SendSnapshotAsync(ws, ct);

            var buffer = new byte[1024];
            while (ws.State == WebSocketState.Open && !ct.IsCancellationRequested)
            {
                var result = await ws.ReceiveAsync(buffer, ct);
                if (result.MessageType == WebSocketMessageType.Close)
                    break;
            }
        }
        finally
        {
            _clients.TryRemove(clientId, out _);
            if (ws.State == WebSocketState.Open)
                await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, null, CancellationToken.None);
        }
    }

    private async Task SendSnapshotAsync(WebSocket ws, CancellationToken ct)
    {
        var snapshot = new { type = "snapshot", trains = _trains.Values.ToArray() };
        await SendJsonAsync(ws, snapshot, ct);
    }

    private async Task BroadcastUpdateAsync(TrainPosition position)
    {
        var msg = new { type = "update", train = position };
        await BroadcastAsync(msg);
    }

    private async Task BroadcastRemovalAsync(string rid)
    {
        var msg = new { type = "remove", rid };
        await BroadcastAsync(msg);
    }

    private async Task BroadcastAsync(object message)
    {
        var dead = new List<string>();

        foreach (var (id, ws) in _clients)
        {
            if (ws.State != WebSocketState.Open)
            {
                dead.Add(id);
                continue;
            }
            try
            {
                await SendJsonAsync(ws, message, CancellationToken.None);
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Failed to send to client {Id}", id);
                dead.Add(id);
            }
        }

        foreach (var id in dead)
            _clients.TryRemove(id, out _);
    }

    private static async Task SendJsonAsync(WebSocket ws, object message, CancellationToken ct)
    {
        var json = JsonSerializer.Serialize(message, new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
        });
        var bytes = Encoding.UTF8.GetBytes(json);
        await ws.SendAsync(bytes, WebSocketMessageType.Text, true, ct);
    }
}

File: frontend/lib/services/train-state-service.ts
import { WebSocket, WebSocketServer } from 'ws'
import type { IncomingMessage } from 'http'
import type { TrainPosition } from '../../types/train'

export class TrainStateService {
  private readonly trains = new Map<string, TrainPosition>()
  private readonly clients = new Map<string, WebSocket>()

  attach(wss: WebSocketServer) {
    wss.on('connection', (ws: WebSocket, _req: IncomingMessage) => {
      const id = Math.random().toString(36).slice(2)
      this.clients.set(id, ws)

      this.sendSnapshot(ws)

      ws.on('close', () => this.clients.delete(id))
      ws.on('error', () => {
        this.clients.delete(id)
        ws.terminate()
      })
    })
  }

  updateTrain(position: TrainPosition) {
    this.trains.set(position.rid, position)
    this.broadcast({ type: 'update', train: position })
  }

  removeTrain(rid: string) {
    this.trains.delete(rid)
    this.broadcast({ type: 'remove', rid })
  }

  getSnapshot(): TrainPosition[] {
    return Array.from(this.trains.values())
  }

  private sendSnapshot(ws: WebSocket) {
    this.send(ws, { type: 'snapshot', trains: Array.from(this.trains.values()) })
  }

  private broadcast(msg: object) {
    for (const [id, ws] of this.clients) {
      if (ws.readyState !== WebSocket.OPEN) {
        this.clients.delete(id)
        continue
      }
      this.send(ws, msg)
    }
  }

  private send(ws: WebSocket, msg: object) {
    try {
      ws.send(JSON.stringify(msg))
    } catch {
      // client gone
    }
  }
}
